In [6]:
import pandas as pd 
import numpy as np

In [8]:
pd.__version__

'1.5.3'

In [7]:
url = '../data/housing.csv'
df = pd.read_csv(url)

In [3]:
df.head(2)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


In [22]:
df.describe().round(2)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
count   20640.00  20640.00            20640.00     20640.00        20433.00   
mean     -119.57     35.63               28.64      2635.76          537.87   
std         2.00      2.14               12.59      2181.62          421.39   
min      -124.35     32.54                1.00         2.00            1.00   
25%      -121.80     33.93               18.00      1447.75          296.00   
50%      -118.49     34.26               29.00      2127.00          435.00   
75%      -118.01     37.71               37.00      3148.00          647.00   
max      -114.31     41.95               52.00     39320.00         6445.00   

       population  households  median_income  median_house_value  
count    20640.00    20640.00       20640.00            20640.00  
mean      1425.48      499.54           3.87           206855.82  
std       1132.46      382.33           1.90           115395.62  
min          3.00        1.00           0.50            14999.00  
25%        787.00      280.00           2.56           119600.00  
50%       1166.00      409.00           3.53           179700.00  
75%       1725.00      605.00           4.74           264725.00  
max      35682.00     6082.00          15.00           500001.00

In [10]:
df.shape

(20640, 10)

In [11]:
df.isnull().sum()

longitude               0
latitude                0
housing_median_age      0
total_rooms             0
total_bedrooms        207
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

In [12]:
df.ocean_proximity.nunique()

5

In [15]:
df.ocean_proximity.unique()

array(['NEAR BAY', '<1H OCEAN', 'INLAND', 'NEAR OCEAN', 'ISLAND'],
      dtype=object)

In [21]:
df.query('ocean_proximity == "NEAR BAY"').median_house_value.mean()

259212.31179039303

In [20]:
df.columns

Index(['longitude', 'latitude', 'housing_median_age', 'total_rooms',
       'total_bedrooms', 'population', 'households', 'median_income',
       'median_house_value', 'ocean_proximity'],
      dtype='object')

### Q.6

1. Calculate the average of `total_bedrooms` column in the dataset.
2. Use the `fillna` method to fill the missing values in `total_bedrooms` with the `mean` value from the previous step.
3. Now, calculate the average of `total_bedrooms` again.
4. Has it changed?

In [45]:
df.total_bedrooms.mean().round(3)

537.871

In [46]:
avg1 = df.total_bedrooms.mean().round(3)

In [27]:
df.total_bedrooms.fillna('mean').isna().sum()

0

In [36]:
i = df[df.total_bedrooms.isna()].index

In [37]:
df.total_bedrooms.fillna(df.total_bedrooms.mean().round(3)).loc[i]

290      537.871
341      537.871
538      537.871
563      537.871
696      537.871
          ...   
20267    537.871
20268    537.871
20372    537.871
20460    537.871
20484    537.871
Name: total_bedrooms, Length: 207, dtype: float64

In [40]:
# doesn't work if fillna('mean')
df.total_bedrooms.fillna(np.mean(df.total_bedrooms).round(3)).loc[i]

290      537.871
341      537.871
538      537.871
563      537.871
696      537.871
          ...   
20267    537.871
20268    537.871
20372    537.871
20460    537.871
20484    537.871
Name: total_bedrooms, Length: 207, dtype: float64

In [41]:
df.total_bedrooms = df.total_bedrooms.fillna(np.mean(df.total_bedrooms).round(3))

In [43]:
df.total_bedrooms.mean().round(3)

537.8705570251938

In [47]:
avg2 = df.total_bedrooms.mean().round(3)

In [48]:
avg1 == avg2

True

### Q.7
1. Select all the options located on `islands`.

In [49]:
df.query('ocean_proximity == "ISLAND"')

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
8314    -118.32     33.35                27.0       1675.0           521.0   
8315    -118.33     33.34                52.0       2359.0           591.0   
8316    -118.32     33.33                52.0       2127.0           512.0   
8317    -118.32     33.34                52.0        996.0           264.0   
8318    -118.48     33.43                29.0        716.0           214.0   

      population  households  median_income  median_house_value  \
8314       744.0       331.0         2.1579            450000.0   
8315      1100.0       431.0         2.8333            414700.0   
8316       733.0       288.0         3.3906            300000.0   
8317       341.0       160.0         2.7361            450000.0   
8318       422.0       173.0         2.6042            287500.0   

     ocean_proximity  
8314          ISLAND  
8315          ISLAND  
8316          ISLAND  
8317          ISLAND  
8318          ISLAND

In [51]:
islands = df[df.ocean_proximity == 'ISLAND']

2. Select only columns `housing_median_age`, `total_rooms`, `total_bedrooms`.

In [52]:
islands.loc[:, ['housing_median_age', 'total_rooms', 'total_bedrooms']]

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [53]:
islands = islands.loc[:, ['housing_median_age', 'total_rooms', 'total_bedrooms']]

3. Get the underlying NumPy array. Let's call it `X`.

In [54]:
islands.values

array([[  27., 1675.,  521.],
       [  52., 2359.,  591.],
       [  52., 2127.,  512.],
       [  52.,  996.,  264.],
       [  29.,  716.,  214.]])

In [55]:
X = islands.values

In [70]:
X.shape

(5, 3)

4. Compute matrix-matrix multiplication between the __transpose of X__ and __X__. To get the transpose, use `X.T`. Let's call the result `XTX`.

In [66]:
X.T.dot(X)

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [62]:
import linalg as la
from importlib import reload

In [69]:
la.matrix_matrix_multiplication(X.T, X)

array([[9.6820000e+03, 3.5105300e+05, 9.1357000e+04],
       [3.5105300e+05, 1.4399307e+07, 3.7720360e+06],
       [9.1357000e+04, 3.7720360e+06, 9.9835800e+05]])

In [71]:
XTX = X.T.dot(X)

5. Compute the inverse of `XTX`.

In [73]:
np.linalg.inv(XTX)

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [76]:
XTX_inv = np.linalg.inv(XTX)

6. Create an array `y` with values `[950, 1300, 800, 1000, 1300]`.

In [74]:
y = np.array([950, 1300, 800, 1000, 1300])

In [75]:
y

array([ 950, 1300,  800, 1000, 1300])

7. Multiply the inverse of `XTX` with the *transpose* of `X`, and then multiply the result by `y`. Call the result `w`.

In [80]:
XTX_inv.dot(X.T).dot(y)

array([23.12330961, -1.48124183,  5.69922946])

8. What's the value of the last element of `w`? - `5.69923`